### PR TITLE
feat: Social logins

### LINKED ISSUE #23: As a user I would like to be able to login/sign-up using a social account, so I don't need to create a profile from scratch

### PR DESCRIPTION
**Pull request by yugantarjain**
### Description
This PR implements social logins on the app for two providers - Apple and Google. Now, a user can sign into the app using username/password, or through the convenient social sign-in methods.
The social sign-in process uses new callback APIs on the backend ([PR](https://github.com/anitab-org/mentorship-backend/pull/719)).

To test - 
1. The backend server must be setup on the local machine and the above mentioned PR shall be used.
2. Apple Developer Account must be used in Xcode to create certificates and 'Sign in with Apple' capability shall be added.

Note: run 'pod install' in the project directory before running and use .xcworkspace instead of .xcodeproj
(Instructions for setting up the project in the README have been updated)

### Mocks
| Login Screen | Apple Sign In | Google Sign In |
| ------------- | ------------- | --------------- |
image | Screen Shot 2020-08-14 at 5 37 28 AM | Screen Shot 2020-08-14 at 5 37 54 AM

Fixes #23 

### Type of Change:
- Code
- Quality Assurance
- User Interface
- Outreach
- Documentation

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Tested on Simluator

### Checklist:
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged 

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by sunjunkie** [comment]
My comments so far:
1 - Social logins should be visible on the sign up and login screens without scrolling. You will have to adjust these screen layouts. (tested on iPhone 8)
2 - Need to note in the docs that to test social logins you must set the developer account AND change the bundle identifier.

**Comment by yugantarjain** [comment]
> My comments so far:
> 1 - Social logins should be visible on the sign up and login screens without scrolling. You will have to adjust these screen layouts. (tested on iPhone 8)
> 2 - Need to note in the docs that to test social logins you must set the developer account AND change the bundle identifier.

1. I think it'd be better to have them only on the login screen and not sign-up (since the backend API handles both the cases), otherwise the user may get a wrong idea about which button on which screen does what. To show them without scrolling, one solution can be to remove logo image and use a navigation bar instead in login screen...
What do you think?

2. Okay.

**Comment by sunjunkie** [comment]
1- yes I think you are right about the signup screen.

I can’t envision what you mean by the nav bar. But the issue only occurs on certain smaller screens.

**Comment by yugantarjain** [comment]
@sunjunkie to show sign-in buttons without scrolling on small screens, I can think of two solutions - 

1. Remove scroll view. SwiftUI manages to show everything by scaling down the mentorship system image (it is resizable, image size now depends on the screen size).

2. Remove image and use a simple title instead (like we do in sign-up screen).

| Method 1 | Method 2 |
| --------- | ---------- |
IMG_EA267E583F59-1 | IMG_A2285924F989-1

Personally, I think the current implementation (scroll view) works well. However, if we do want to remove that, I think Method 2 looks better.

**Comment by sunjunkie** [comment]
Apple's guidelines specifically say to avoid scrolling. "Prominently display a Sign in with Apple button. Make a Sign in with Apple button no smaller than other sign-in buttons, and avoid making people scroll to see the button." https://developer.apple.com/design/human-interface-guidelines/sign-in-with-apple/overview/buttons/

As for the 2 alternatives, the first is my choice. Just alter the image size. I expect this screen design will be decided by the UI team eventually anyway.

**Comment by yugantarjain** [comment]
> My comments so far:
> 1 - Social logins should be visible on the sign up and login screens without scrolling. You will have to adjust these screen layouts. (tested on iPhone 8)
> 2 - Need to note in the docs that to test social logins you must set the developer account AND change the bundle identifier.

1. Done
2. Where exactly should I add it, in the docs folder (in contributing and developing) OR in the readme?

**Comment by vatsalkul** [comment]
**UI Issues**

Signin button are looking good in dark mode



When there is an error message. Google sign in button slides down and break the bounds of screen. Also logo disappeared 



Suggestions: 
1. Decrease size of username and password field heights. (According to screen) 
2. Invert color of apple sign In button to white when device is in dark mode.
3. Make sure logo is visible on the home screen.

File: mentorship ios/AppDelegate.swift [hunk 1]
```diff
@@ -4,33 +4,61 @@
...
-class AppDelegate: UIResponder, UIApplicationDelegate {
-
+class AppDelegate: UIResponder, UIApplicationDelegate, GIDSignInDelegate {
+    let loginService: LoginService = LoginAPI()
+    var loginViewModel = LoginViewModel()
+    
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
-        // Override point for customization after application launch.
+        // Initialize sign-in
+        GIDSignIn.sharedInstance().clientID = "992237180107-lsibe891591qcubpbd8qom4fts74i5in.apps.googleusercontent.com"
```
Comment (ApheleiaS) [inline mentorship ios/AppDelegate.swift]: Client ID secret in app code, can we move it elsewhere?

File: mentorship ios/Info.plist [hunk 2]
```diff
@@ -16,6 +16,17 @@
...
 	<key>CFBundleShortVersionString</key>
 	<string>1.0</string>
+	<key>CFBundleURLTypes</key>
+	<array>
+		<dict>
+			<key>CFBundleTypeRole</key>
+			<string>Editor</string>
+			<key>CFBundleURLSchemes</key>
+			<array>
+				<string>com.googleusercontent.apps.992237180107-lsibe891591qcubpbd8qom4fts74i5in</string>
```
Comment (ApheleiaS) [inline mentorship ios/Info.plist]: Same here.

File: mentorship ios/Views/Registration/Login.swift [hunk 3]
```diff
@@ -68,9 +59,34 @@ struct Login: View {
...
+            //activity indicator or show user message text
+            if self.loginViewModel.inActivity {
+                ActivityIndicator(isAnimating: $loginViewModel.inActivity)
+            } else if !(self.loginViewModel.loginResponseData.message?.isEmpty ?? true) {
+                Text(self.loginViewModel.loginResponseData.message ?? "")
+                    .modifier(ErrorText())
+            }
+            
+            // Divider for social sign in options
+            ZStack {
```
Comment (ApheleiaS) [inline mentorship ios/Views/Registration/Login.swift]: Shouldn't this be in Vstack?

Comment (yugantarjain) [reply mentorship ios/Views/Registration/Login.swift]: This ZStack (used to shown "Or" text on top of divider) is inside VStack.

File: mentorship ios/Info.plist (hunk 2 above)
Comment (yugantarjain) [reply mentorship ios/Info.plist]: This I guess can't be altered. For google sign in, we add reversed client id as url type, xcode automatically adds it to plist.

**Comment by yugantarjain** [comment]
@vatsalkul UI issues fixed.

File: mentorship ios/AppDelegate.swift (hunk 1 above)
Comment (yugantarjain) [reply mentorship ios/AppDelegate.swift]: Done. Moved to a new Config.plist file.

**Comment by yugantarjain** [comment]
@vatsalkul social sign in buttons made into one line. Also, space between buttons and divider reduced.

**Comment by sunjunkie** [comment]
@yugantarjain This looks good. Update the base branch first and I will approve.

**Comment by yugantarjain** [comment]
@sunjunkie updated.
@anitab-org/mentorship-ios-maintainers have updated the docs too and have removed the google auth client id from app.

**Review by sunjunkie — APPROVED** [review]

**Review by vatsalkul — APPROVED** [review]

**Pull request merged**